digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/NodesAndJunctionsService.scala

```diff
@@ -496,90 +497,86 @@ class NodesAndJunctionsService(roadwayDAO: RoadwayDAO, roadwayPointDAO: RoadwayP
           roadTypeSections.foreach { section =>
 
-            val headLink = section.head
-            val lastLink = section.last
-
-            val headRoadwayPointId = {
-              val existingRoadwayPoint = roadwayPointDAO.fetch(headLink.roadwayNumber, headLink.startAddrMValue)
-              if (existingRoadwayPoint.nonEmpty)
-                existingRoadwayPoint.get.id
-              else roadwayPointDAO.create(headLink.roadwayNumber, headLink.startAddrMValue, username)
-            }
-            val lastRoadwayPointId = {
-              val existingRoadwayPoint = roadwayPointDAO.fetch(lastLink.roadwayNumber, lastLink.endAddrMValue)
-              if (existingRoadwayPoint.nonEmpty)
-                existingRoadwayPoint.get.id
-              else roadwayPointDAO.create(lastLink.roadwayNumber, lastLink.endAddrMValue, username)
-            }
-
-            /*  Handle update of NODE_POINT in reverse cases  */
-            val (startNodeReversed, endNodeReversed) =
-              (roadwayChanges.exists(ch =>
-                ch.changeInfo.target.startAddressM.nonEmpty && headLink.startAddrMValue == ch.changeInfo.target.startAddressM.get && ch.changeInfo.reversed
-              ),
-                roadwayChanges.exists(ch =>
-                  ch.changeInfo.target.endAddressM.nonEmpty && lastLink.endAddrMValue == ch.changeInfo.target.endAddressM.get && ch.changeInfo.reversed
-                ))
-
-
-            val existingHeadNodePoint = {
-              val originalLink = mappedRoadwayNumbers.find(mpr => headLink.startAddrMValue == mpr.newStartAddr && headLink.endAddrMValue == mpr.newEndAddr && mpr.newRoadwayNumber == headLink.roadwayNumber)
-              if (originalLink.nonEmpty) {
-                if (!startNodeReversed) {
-                  nodePointDAO.fetchNodePointsTemplates(Set(originalLink.get.originalRoadwayNumber, headLink.roadwayNumber)).find(np => np.beforeAfter == After && np.addrM == headLink.startAddrMValue)
-                } else {
-                  nodePointDAO.fetchNodePointsTemplates(Set(originalLink.get.originalRoadwayNumber, headLink.roadwayNumber)).find(np => np.beforeAfter == Before && np.addrM == originalLink.get.newStartAddr)
-                }
-              } else None
+            val headProjectLink = section.head
+            val headReversed = roadwayChanges.exists(ch => ch.changeInfo.target.startAddressM.nonEmpty && headProjectLink.startAddrMValue == ch.changeInfo.target.startAddressM.get && ch.changeInfo.reversed)
+
+            val headNodePoint = mappedRoadwayNumbers.find { rl =>
+              headProjectLink.startAddrMValue == rl.newStartAddr && headProjectLink.endAddrMValue == rl.newEndAddr && headProjectLink.roadwayNumber == rl.newRoadwayNumber
+            }.flatMap { rl =>
+              if (headReversed) {
+                nodePointDAO.fetchByRoadwayNumbers(Seq(rl.originalRoadwayNumber, headProjectLink.roadwayNumber))
```

Comment:
nodePointDAO.fetchByRoadwayNumbers is used 4 times
I think nodePointDAO.fetchByRoadwayNumbers returns too many nodepoints
It returns calculated nodepoints and I think it should not do that. 
It uses left join to node which means that there is no need to node exist at all so it might return oprhaned nodepoints that are not connected to node.

Reply:
nodePointDAO.fetchByRoadwayNumbers should return every node points (templates, calculated, road) so we know if we need to create a new one or not.

Basically if a node needs to be modified - due to:
1. Roadway point change
2. Before/After
3. Type

It's possible that the node doesn't exist yet, so it's a simple create, but if the node exists then we need to know it to expire it and create new one.

I tried my best to improve that code, before it was way much more raw code.
Although the query is written 4 times, it will only be called twice. One for the 1st point of the section, and another one for the last.

The main problem improving this code, is that in reverse cases, the address change and the beforeAfter is also reversed. Which makes it more complicated to found the right node.

Reply:
See other comment why calculated node point should not be fetched.